Added example with Workflow interface for finetuning Llama2 LLM

manuelhsantana opened the pull request:
This PR introduces a new example of fine-tuning the Llama2 Language Model (LLM), using the workflow interface. 

The main objective of this PR is to provide users and developers with a practical guide on how to fine-tune the Llama2 LLM for their specific use cases.

The added example includes:

- Steps on how to load the Llama2 LLM and prepare it for fine-tuning.
- Instructions on how to set up the training and validation datasets.
- Use the OpenFL workflow interface for fine-tune the model with a specific dataset.

A few things to keep in mind:
- For this new example it is a prerequisite to register and request access to the Meta model.
- The lvwerra and some Huggingface libraries are used.
 
This tutorial serves as a basic example, and users are encouraged to adapt and expand upon it to suit their specific needs and requirements.

Users are animated to explore this example and provide feedback, which will be invaluable in refining and expanding our set of examples.

Please review the changes and provide your valuable feedback.

manuelhsantana committed:
Added example for finetuning Llama2 LLM

manuelhsantana committed:
Update notebook to use gpu

manuelhsantana committed:
Updated ipynb file to remove minor typos.

psfoley commented on openfl-tutorials/experimental/Workflow_Interface_501_FineTuning_LLAMA2.ipynb:
```diff
@@ -0,0 +1,874 @@
+   "metadata": {},
+   "source": [
+    "Welcome to the introductory tutorial for the OpenFL Experimental Workflow Interface! This notebook is designed to guide you through the process of setting up your first horizontal federated learning workflow using the pre-trained Large Language Model Llama-2. The primary objectives of this tutorial are:\r\n",
+    "- Simplify the federated workflow representatio.n\r\n",
```
```suggestion
    "- Simplify the federated workflow representation\r\n",
```

manuelhsantana replied:
Fixed.

psfoley commented on openfl-tutorials/experimental/Workflow_Interface_501_FineTuning_LLAMA2.ipynb:
```diff
@@ -0,0 +1,874 @@
+   "source": [
+    "Welcome to the introductory tutorial for the OpenFL Experimental Workflow Interface! This notebook is designed to guide you through the process of setting up your first horizontal federated learning workflow using the pre-trained Large Language Model Llama-2. The primary objectives of this tutorial are:\r\n",
+    "- Simplify the federated workflow representatio.n\r\n",
+    "- Help users better understand the steps in federated learni.ng\r\n",
```
```suggestion
    "- Help users better understand the steps in federated learning\r\n",
```

manuelhsantana replied:
Fixed.

psfoley commented on openfl-tutorials/experimental/Workflow_Interface_501_FineTuning_LLAMA2.ipynb:
```diff
@@ -0,0 +1,874 @@
+    "            packing=False,\n",
+    "            dataset_text_field=\"text\",\n",
+    "            tokenizer=tokenizer, \n",
+    "            max_seq_length=128,#64\n",
```
Remove old max_seq_length

manuelhsantana replied:
Fixed.

psfoley reviewed (commented):
Thanks @manuelhsantana. This PR is very close. Three small changes to address, then this is ready to merge.

manuelhsantana committed:
Update Workflow_Interface_501_FineTuning_LLAMA2.ipynb

Fix typos and unused comments

manuelhsantana committed:
Fix typos Workflow_Interface_501_FineTuning_LLAMA2.ipynb

kminhta commented on openfl-tutorials/experimental/Workflow_Interface_501_FineTuning_LLAMA2.ipynb:
```diff
@@ -0,0 +1,874 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -r requirements_workflow_interface.txt"
```
duplicate command 
`!pip install -r requirements_workflow_interface.txt` already in previous cell

manuelhsantana replied:
Fixed

kta-intel commented:
I am trying to run it on CPU, but am running error:
`NameError: name 'str2optimizer32bit' is not defined`

@manuelhsantana did you happen to run into this? I can try to triage a bit more, but from a quick investigate, it seems like it is expecting CUDA drivers, which I think is a req for bitesandbytes currently

for easier testing, I reduced dataset size by replacing the data cell with this:
```
dataset=load_dataset(dataset_name,split='train[10:20]')

dataset = DatasetDict({
    'train': dataset,
    'test': dataset,
    'valid': dataset})
```

manuelhsantana committed:
Added hash validation to dataset example

psfoley commented on openfl-tutorials/experimental/Workflow_Interface_501_FineTuning_LLAMA2.ipynb:
```diff
@@ -69,8 +69,7 @@
    "source": [
-    "!pip install git+https://github.com/intel/openfl.git\n",
-    "!pip install -r requirements_workflow_interface.txt"
+    "!pip install git+https://github.com/intel/openfl.git"
```
Reminder to change this to `https://github.com/securefederatedai/openfl.git`

manuelhsantana replied:
Fixed

manuelhsantana committed:
Fixed str2optimizer32bit warning on CPU run

manuelhsantana commented:
> I am trying to run it on CPU, but am running error: `NameError: name 'str2optimizer32bit' is not defined`
> 
> @manuelhsantana did you happen to run into this? I can try to triage a bit more, but from a quick investigate, it seems like it is expecting CUDA drivers, which I think is a req for bitesandbytes currently
> 
> for easier testing, I reduced dataset size by replacing the data cell with this:
> 
> ```
> dataset=load_dataset(dataset_name,split='train[10:20]')
> 
> dataset = DatasetDict({
>     'train': dataset,
>     'test': dataset,
>     'valid': dataset})
> ```

I changed the optimizer to avoid the warning on CPU

kta-intel commented:
> > I am trying to run it on CPU, but am running error: `NameError: name 'str2optimizer32bit' is not defined`
> > @manuelhsantana did you happen to run into this? I can try to triage a bit more, but from a quick investigate, it seems like it is expecting CUDA drivers, which I think is a req for bitesandbytes currently
> > for easier testing, I reduced dataset size by replacing the data cell with this:
> > ```
> > dataset=load_dataset(dataset_name,split='train[10:20]')
> > 
> > dataset = DatasetDict({
> >     'train': dataset,
> >     'test': dataset,
> >     'valid': dataset})
> > ```
> 
> I changed the optimizer to avoid the warning on CPU

Thanks - this resolved the issue on my end. Everything looks good to me. Great contribution!

rahulga1 commented:
@manuelhsantana can you fix the pipeline failures and ask for re-review again?

Commit:
fix code format

rajithkrishnegowda commented:
> @manuelhsantana can you fix the pipeline failures and ask for re-review again?

resolved the pipeline failures

rahulga1 commented:
@kta-intel @psfoley Rajith has fixed pipeline issues, can you review this, if it looks ok, lets merge it?

kta-intel commented:
> @kta-intel @psfoley Rajith has fixed pipeline issues, can you review this, if it looks ok, lets merge it?

This seems overall similar with the [Phi-4 example](https://github.com/securefederatedai/openfl/blob/develop/openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb). Given that llama-2 is also already an outdated model, unless we update the model and highlight some new feature differences, I suggest we close this.

@manuelhsantana if you still wish to add it, I invite you to reopen this in the openfl-contrib.
